### PR TITLE
[#1053] Adding more credentials tests

### PR DESCRIPTION
**Pull request by rsandell**
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by rsandell** [comment]
Putting it in draft because I have more related tests to add and I haven't gotten all these to succeed locally yet.

**Comment by rsandell** [comment]
Now I think I need some help figuring out what is going on. Because the tests don't fail locally for me any more.

File: src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/AbstractCredentialsTest.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        security.save();
+
+        //Make sure we have login successfully before continuing
+        assertLogin(Matchers.loggedInAs(CRED_USER));
+    }
+
+    private void assertLogin(Matcher<Login> matcher) {
+        Login login = jenkins.login();
+        login.doLogin(CRED_USER);
+        //Give some time to login to do backgroun proccesses (found some times when the user page shows you are not logged even if the link is present)
```
Comment (car-roll) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/AbstractCredentialsTest.java:74]:
```suggestion
        //Give some time to login to do background proccesses (found some times when the user page shows you are not logged even if the link is present)
```

File: src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/AbstractCredentialsTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import org.jenkinsci.test.acceptance.plugins.ssh_credentials.SshPrivateKeyCredential;
+import org.jenkinsci.test.acceptance.po.GlobalSecurityConfig;
+import org.jenkinsci.test.acceptance.po.Login;
+import org.jenkinsci.test.acceptance.selenium.UselessFileDetectorReplacement;
+import org.openqa.selenium.WebElement;
+
+import java.io.File;
+import java.util.Locale;
+
+import static org.junit.Assert.assertThat;
```
Comment (NotMyFault) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/AbstractCredentialsTest.java:39]:
```suggestion
import static org.hamcrest.MatcherAssert.assertThat;
```

File: src/test/java/plugins/CredentialsBindingTest.java [hunk 3]
```diff
@@ -45,4 +96,113 @@ public void testTextBinding() {
         b.shouldSucceed();
         assertThat(b.getConsole(), containsString(SECRET_OUTPUT));
     }
+
+    @Test
+    @WithPlugins("credentials-binding@1.10")
+    public void pipelineWithCredentialsTest() throws IOException {
+        final HttpResponse resp = new CredentialsRESTClient(jenkins.url).createCredential(CRED_ID, CRED_USER, CRED_PWD, GLOBAL_SCOPE);
+        Assert.assertThat(resp.getStatusLine().getStatusCode(), is(200));
```
Comment (NotMyFault) [inline src/test/java/plugins/CredentialsBindingTest.java:104]:
```suggestion
       assertThat(resp.getStatusLine().getStatusCode(), is(200));
```

File: src/test/java/plugins/CredentialsBindingTest.java [hunk 4]
```diff
@@ -45,4 +96,113 @@ public void testTextBinding() {
...
+        } else {
+            b.shouldFail();
+        }
+
+        //Job ending does not imply that console log has been completely updated so we wait
+        waitForLogToBeFullyLoaded(b, "Finished");
+
+        final String consoleOutput = b.getConsole();
+        for (final String message : messagesToCheck) {
+            Assert.assertThat(consoleOutput, Matchers.containsString(message));
```
Comment (NotMyFault) [inline src/test/java/plugins/CredentialsBindingTest.java:175]:
```suggestion
           assertThat(consoleOutput, Matchers.containsString(message));
```

**Review by NotMyFault — APPROVED** [review]
Thanks, I left some feedback for consideration, based on a quick read.

File: src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/CredentialsRESTClient.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    private static final String CONTENT_TYPE_HEADER_NAME = "Content-Type";
+    private static final String CONTENT_TYPE = "application/xml;charset=utf-8";
+
+    private final HttpClient httpClient;
+    private final URL jenkinsUrl;
+
+
+    public CredentialsRESTClient(final URL jenkinsUrl) {
+        this.jenkinsUrl = jenkinsUrl;
+        this.httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build();
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/CredentialsRESTClient.java:62]: there is functionality somewhere to get a rest client that is configured with authentication.

https://github.com/jenkinsci/acceptance-test-harness/blob/d008ba64246853c1aadd564e93523528a540d1cd/src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java#L306-L318

File: src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/CredentialsRESTClient.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    private static final String CREATE_DOMAINS_URL = BASE_REST_URL + "/createDomain";
+
+    private static final String CONTENT_TYPE_HEADER_NAME = "Content-Type";
+    private static final String CONTENT_TYPE = "application/xml;charset=utf-8";
+
+    private final HttpClient httpClient;
+    private final URL jenkinsUrl;
+
+
+    public CredentialsRESTClient(final URL jenkinsUrl) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/CredentialsRESTClient.java:60]: I am wondering why we should be doing this via the REST API and not via the Jenkins UI?
if we want to test the RestAPI we can do that inside the credentials plugin itself in a `RealJenkinsRule`

Comment (rsandell) [reply src/main/java/org/jenkinsci/test/acceptance/plugins/credentials/CredentialsRESTClient.java:60]: Yea, it is weird that it is not using the UI to create the credentials, the only explanation I can come up with for now  is speed.

**Review by julieheard — APPROVED** [review]

**Comment by raul-arabaolaza** [comment]
Remaining failing tests are flaky, let's give a try to fix them before merging or retriggering this one, I am gonna add my debugging code from https://github.com/jenkinsci/acceptance-test-harness/pull/1054/files to see if we can reproduce or even fix this

**Comment by timja** [comment]
> Remaining failing tests are flaky, let's give a try to fix them before merging or retriggering this one, I am gonna add my debugging code from [#1054 (files)](https://github.com/jenkinsci/acceptance-test-harness/pull/1054/files) to see if we can reproduce or even fix this

It would be great to fix them, you could also disable them and file issues for them as well.

File: src/main/java/org/jenkinsci/test/acceptance/po/WizardCreateAdminUser.java [hunk 7]
```diff
@@ -74,9 +78,18 @@ public void shouldCreateTheUserSuccessfully() {
 
     public void confirmURLSettings(){
         By confirmJenkinsUrl = by.css(".btn-primary.save-configure-instance");
-        waitFor(confirmJenkinsUrl,2);
+        waitFor(driver).withTimeout(Duration.ofSeconds(time.seconds(2)))
```
Comment (raul-arabaolaza) [inline src/main/java/org/jenkinsci/test/acceptance/po/WizardCreateAdminUser.java:81]: Unrelated to this change, but should help on diagnosing `InstallWizardTest` failures if they happen again

**Review by raul-arabaolaza — APPROVED** [review]

**Pull request merged**